File: models/traditional.py
"""Black-Scholes and Binomial option pricers (closed-form, no training)."""

import numpy as np
from scipy.stats import norm
from scipy.optimize import brentq

from models.base import OptionPricer


class BlackScholesPricer(OptionPricer):
    """Black-Scholes closed-form pricer for European calls.

    Always available, no training required.
    """

    def __init__(self, name: str = "bs"):
        super().__init__(name)

    def _d1(self, S, K, T, r, q, sigma=0.25):
        return (np.log(S / K) + (r - q + 0.5 * sigma ** 2) * T) / (sigma * np.sqrt(T))

    def _d2(self, d1, T, sigma=0.25):
        return d1 - sigma * np.sqrt(T)

    def price(self, S, K, T, r=0.045, q=0, sigma=0.25):
        """European call price."""
        if T <= 0:
            return max(float(S - K), 0.0)
        d1 = self._d1(S, K, T, r, q, sigma)
        d2 = self._d2(d1, T, sigma)
        return float(S * np.exp(-q * T) * norm.cdf(d1) - K * np.exp(-r * T) * norm.cdf(d2))

    def greeks(self, S, K, T, r=0.045, q=0, sigma=0.25):
        """Compute Greeks for a European call."""
        if T <= 0:
            return {"delta": float(S > K), "gamma": 0.0, "vega": 0.0, "theta": 0.0}
        d1 = self._d1(S, K, T, r, q, sigma)
        d2 = self._d2(d1, T, sigma)
        delta = np.exp(-q * T) * norm.cdf(d1)
        gamma = np.exp(-q * T) * norm.pdf(d1) / (S * sigma * np.sqrt(T))
        vega = S * np.exp(-q * T) * norm.pdf(d1) * np.sqrt(T) / 100  # per 1% vol move
        theta = (-(S * np.exp(-q * T) * norm.pdf(d1) * sigma) / (2 * np.sqrt(T))
                 - r * K * np.exp(-r * T) * norm.cdf(d2)
                 + q * S * np.exp(-q * T) * norm.cdf(d1)) / 365
        return {"delta": float(delta), "gamma": float(gamma),
                "vega": float(vega), "theta": float(theta)}

    def surface(self, strikes, expiries, S, r=0.045, q=0, sigma=0.25):
        """Return [NE, NS] price matrix."""
        strikes = np.asarray(strikes, dtype=np.float32)
        expiries_T = np.array([e["T"] if isinstance(e, dict) else e for e in expiries], dtype=np.float32)
        NE, NS = len(expiries_T), len(strikes)
        result = np.zeros((NE, NS), dtype=np.float32)
        for i, T in enumerate(expiries_T):
            for j, K in enumerate(strikes):
                result[i, j] = self.price(S, K, T, r, q, sigma)
        return result

    def implied_vol(self, S, K, T, market_price, r=0.045, q=0):
        """Compute implied vol from a market price using Brent's method."""
        if T <= 0:
            return 0.0
        try:
            def f(sigma):
                return self.price(S, K, T, r, q, sigma) - market_price
            return float(brentq(f, 0.001, 2.0))
        except (ValueError, RuntimeError):
            return 0.25  # fallback to ATM guess

    def attention(self, S, K, T):
        """No attention for closed-form models."""
        return None


class BinomialPricer(OptionPricer):
    """Binomial tree pricer for American options (with early exercise)."""

    def __init__(self, name: str = "binomial", steps: int = 100):
        super().__init__(name)
        self.steps = steps

    def price(self, S, K, T, r=0.045, q=0, sigma=0.25, is_call=True):
        """American option price via binomial tree."""
        if T <= 0:
            return max(float(S - K) if is_call else float(K - S), 0.0)
        dt = T / self.steps
        u = np.exp(sigma * np.sqrt(dt))
        d = 1.0 / u
        p = (np.exp((r - q) * dt) - d) / (u - d)

        # Price at expiry (terminal nodes)
        ST = S * (u ** (np.arange(self.steps + 1))) * (d ** (self.steps - np.arange(self.steps + 1)))
        if is_call:
            V = np.maximum(ST - K, 0)
        else:
            V = np.maximum(K - ST, 0)

        # Backwards induction
        for i in range(self.steps - 1, -1, -1):
            V = np.exp(-r * dt) * (p * V[:i + 1] + (1 - p) * V[1:i + 2])
            # Early exercise check for American
            S_t = S * (u ** np.arange(i + 1)) * (d ** (i - np.arange(i + 1)))
            if is_call:
                V = np.maximum(V, S_t - K)
            else:
                V = np.maximum(V, K - S_t)
        return float(V[0])

    def surface(self, strikes, expiries, S, r=0.045, q=0, sigma=0.25):
        strikes = np.asarray(strikes, dtype=np.float32)
        expiries_T = np.array([e["T"] if isinstance(e, dict) else e for e in expiries], dtype=np.float32)
        NE, NS = len(expiries_T), len(strikes)
        result = np.zeros((NE, NS), dtype=np.float32)
        for i, T in enumerate(expiries_T):
            for j, K in enumerate(strikes):
                result[i, j] = self.price(S, K, T, r, q, sigma)
        return result

File: src/models/traditional.js
// Black-Scholes and Binomial option pricers (closed-form, no training).
// Port of models/traditional.py.

const { OptionPricer } = require('./base');
const { normCdf, normPdf, brentq } = require('../stats');
const { expiryT, zeros2d } = require('../util');

class BlackScholesPricer extends OptionPricer {
  constructor(name = 'bs') {
    super(name);
  }

  _d1(S, K, T, r, q, sigma = 0.25) {
    return (Math.log(S / K) + (r - q + 0.5 * sigma * sigma) * T) / (sigma * Math.sqrt(T));
  }

  _d2(d1, T, sigma = 0.25) {
    return d1 - sigma * Math.sqrt(T);
  }

  price(S, K, T, r = 0.045, q = 0, sigma = 0.25) {
    if (T <= 0) return Math.max(S - K, 0);
    const d1 = this._d1(S, K, T, r, q, sigma);
    const d2 = this._d2(d1, T, sigma);
    return S * Math.exp(-q * T) * normCdf(d1) - K * Math.exp(-r * T) * normCdf(d2);
  }

  greeks(S, K, T, r = 0.045, q = 0, sigma = 0.25) {
    if (T <= 0) return { delta: S > K ? 1 : 0, gamma: 0, vega: 0, theta: 0 };
    const d1 = this._d1(S, K, T, r, q, sigma);
    const d2 = this._d2(d1, T, sigma);
    const delta = Math.exp(-q * T) * normCdf(d1);
    const gamma = (Math.exp(-q * T) * normPdf(d1)) / (S * sigma * Math.sqrt(T));
    const vega = (S * Math.exp(-q * T) * normPdf(d1) * Math.sqrt(T)) / 100; // per 1% vol move
    const theta =
      (-(S * Math.exp(-q * T) * normPdf(d1) * sigma) / (2 * Math.sqrt(T)) -
        r * K * Math.exp(-r * T) * normCdf(d2) +
        q * S * Math.exp(-q * T) * normCdf(d1)) /
      365;
    return { delta, gamma, vega, theta };
  }

  surface(strikes, expiries, S, r = 0.045, q = 0, sigma = 0.25) {
    const exp = expiries.map(expiryT);
    const NE = exp.length;
    const NS = strikes.length;
    const result = zeros2d(NE, NS);
    for (let i = 0; i < NE; i++) {
      for (let j = 0; j < NS; j++) {
        result[i][j] = this.price(S, strikes[j], exp[i], r, q, sigma);
      }
    }
    return result;
  }

  impliedVol(S, K, T, marketPrice, r = 0.045, q = 0) {
    if (T <= 0) return 0;
    try {
      return brentq((sigma) => this.price(S, K, T, r, q, sigma) - marketPrice, 0.001, 2.0);
    } catch {
      return 0.25; // fallback to ATM guess
    }
  }

  attention() {
    return null;
  }
}

class BinomialPricer extends OptionPricer {
  constructor(name = 'binomial', steps = 100) {
    super(name);
    this.steps = steps;
  }

  price(S, K, T, r = 0.045, q = 0, sigma = 0.25, isCall = true) {
    if (T <= 0) return Math.max(isCall ? S - K : K - S, 0);
    const n = this.steps;
    const dt = T / n;
    const u = Math.exp(sigma * Math.sqrt(dt));
    const d = 1 / u;
    const p = (Math.exp((r - q) * dt) - d) / (u - d);
    const disc = Math.exp(-r * dt);

    // Terminal node values.
    let V = new Array(n + 1);
    for (let i = 0; i <= n; i++) {
      const ST = S * Math.pow(u, i) * Math.pow(d, n - i);
      V[i] = Math.max(isCall ? ST - K : K - ST, 0);
    }

    // Backward induction with early-exercise (American). Indexing mirrors
    // models/traditional.py exactly (V_new[j] = disc*(p*V[j] + (1-p)*V[j+1]))
    // so JS output is numerically identical to the original.
    for (let step = n - 1; step >= 0; step--) {
      const Vn = new Array(step + 1);
      for (let j = 0; j <= step; j++) {
        const cont = disc * (p * V[j] + (1 - p) * V[j + 1]);
        const St = S * Math.pow(u, j) * Math.pow(d, step - j);
        const exercise = isCall ? St - K : K - St;
        Vn[j] = Math.max(cont, exercise);
      }
      V = Vn;
    }
    return V[0];
  }

  surface(strikes, expiries, S, r = 0.045, q = 0, sigma = 0.25) {
    const exp = expiries.map(expiryT);
    const NE = exp.length;
    const NS = strikes.length;
    const result = zeros2d(NE, NS);
    for (let i = 0; i < NE; i++) {
      for (let j = 0; j < NS; j++) {
        result[i][j] = this.price(S, strikes[j], exp[i], r, q, sigma);
      }
    }
    return result;
  }
}

module.exports = { BlackScholesPricer, BinomialPricer };
